api/src/main/java/com/velocitypowered/api/event/ResultedEvent.java

```diff
@@ -106,4 +107,55 @@ public static ComponentResult denied(@NonNull Component reason) {
             return new ComponentResult(false, reason);
         }
     }
+    /**
+     * Represents an "allowed/allowed with online mode/denied" result with a reason allowed for denial.
+     */
+    class PreLoginComponentResult extends ComponentResult {
+
+        private static final PreLoginComponentResult ALLOWED = new PreLoginComponentResult(true, null);
+        private static final PreLoginComponentResult ALLOWED_ONLINEMODE = new PreLoginComponentResult(true);
+
+        private final boolean onlineMode;
+        
+        /**
+         * Allows to enable a online mode for the player connection, when velocity running in offline mode
+         * Does not have any sense if velocity running in onlineMode;
+         * @param onlineMode if true, offline uuid will be used for player connection to avoid collision
```

Comment:
Huh? Note that if Velocity determines that a player with the same username is online, it disconnects the new connection.

Reply:
On a related note, this should really have a `GameProfile` instead. An empty fake game profile means Velocity should perform regular authentication, otherwise Velocity can proceed to login with the fake profile. (In other words, this means that a basic version of a plugin like SkinsRestorer can work on Velocity simply by checking if the proxy is in offline mode, and if so, modifying the result of the `PreLoginEvent` with the new game profile).

Reply:
I mean that a returned UUID from mojang, can be same with uuid of other player that was generated for offline player. Chance is very small, but..

Reply:
Unlikely:

```
>>> import uuid
>>> uuid.UUID('652a2bc4-e8cd-405d-b7b6-98156ee2dc09').version
4
>>> 
```

UUID version 4 IDs are totally random and the chances of two UUID v4 iDs colliding is so minimal that for all intents an purposes, UUIDs are effectively random.

Reply:
I think it better to use offline uuid, if velocity is running in offline mode.

Reply:
For sanity, yes, we should do that. At the very least though, plugins should be able to inject arbitrary profile properties.

Reply:
Here's a usecase for allowing arbitrary UUIDs. If Mojang login servers go down (this used to happen a lot, it's rare now), under certain circumstances a plugin may wish to bypass authentication and fake the player's UUID and their skin.

Reply:
Why don't we try something else? A new event (`GameProfileRequestEvent`) fired just after the pre-login event and a plugin can set an arbitrary game profile there. It's never invoked in online-mode since it doesn't make sense.